### DISCUSSION TITLE
Mac环境下的配置和环境准备

### DISCUSSION THREAD

**Discussion by zRzRzRzRzRzRzR** [zai-org/ChatGLM3#251] [Q&A] [answered]
在这里提出Mac环境下配置ChatGLM3-6B模型的任何问题，例如：

+ Mac推理速度很慢，有没有什么优化方案
+ 我想知道chatglm有没有针对M3芯片的优化推理方案

**Comment by xxxzsgxxx** [comment]
mac下模型推理ggml/gguf量化是王道，否则速度感人，chatglm.cpp项目直接走起。

> 在这里提出Mac环境下配置ChatGLM3-6B模型的任何问题，例如：
> 
> * Mac推理速度很慢，有没有什么优化方案
> * 我想知道chatglm有没有针对M3芯片的优化推理方案

**Reply by ghost** [reply]
期待高阶功能， function calling, code interpreter ggml/gguf量化还能体验么？

**Comment by ghost** [comment]
M1 pro 16GB内存 平均1分钟大概吐一个字的样子

**Reply by craii** [reply]
M3 24G内存，加载chatglm3-6b模型直接炸显存，无法加载完成怎么处理？
![Snipaste_2023-12-04_17-25-18](https://github.com/THUDM/ChatGLM3/assets/10702100/01ed6890-7143-4a6a-92a5-76bd95690a37)

**Reply by zRzRzRzRzRzRzR** [reply]
emm，你这个记得加上half 不然你载入的是fp32的就要25GB内存了

**Reply by craii** [reply]
谢谢你，后来改用文档里提到的 chatglm.cpp 转 int4 处理了。

**Comment by zRzRzRzRzRzRzR** [comment]
这个速度确实太慢了，我们查一下

**Reply by zRzRzRzRzRzRzR** [reply, marked as answer]
目前最好的方案还是使用类似glm-cpp等工具来推理int4版本的，fp16版本在m1上确实是非常非常卡

**Comment by ghost** [comment]
Apple M2 Pro + 32G 内存可正常运行 chatGLM3-6B，但是有一个严重的问题，就是内存占用只增不减，随着使用，由于内存不断的增加，导致推理速度逐渐变慢，直到撑爆物理内存，开始使用交换分区，这可如何是好（N 卡的就正常，显存是会释放的

**Reply by Ventery** [reply]
你好，在N卡上跑的话你的history输入条数有限制吗？正常来说history越多占用内存就应该越大的呀

**Reply by GuiminZHOU** [reply]
同问，m2max+64G，运行web_demo后，4-5次会话，内存从12G激增至56G，且回复内容陷入死循环，一直返回重复的内容

**Reply by mobguang** [reply]
是的，我发现在Mac(M2 ultra，192GB)上运行vicuna-13b也会出现内存一直上涨而且无法回收的现象。不知道大家有没有办法解决。

**Reply by southkorea2013** [reply]
暂时可以将HISTORY_LEN 设置为
 # 历史对话轮数,如果设置不为0，内存占用会一直飙升
	HISTORY_LEN = 0

**Comment by MoYJiang** [comment]
Apple Mac Pro M3 24G内存 使用官方的composite_demo推理巨慢，简单的'你好'需要花费近30分钟，内存利用率很高达到百分之九十多，而CUP和GPU的利用只有百分之十几到二十。从stream_chat的输出来看，卡在推理计算，每个token输出都显得很困难。

**Comment by Ventery** [comment]
M2 pro 芯片+32G内存 已经成功在GPU上跑成功，速度超级快，已经跟chatgpt3.5一样快了，不需要加速
准备工作就是先部署好，可以生成回答的状态（虽然超级慢）
然后按照以下步骤来
1、先删掉这哥仨
`pip uninstall torch torchvision torchaudio
`
2、从官网上下nightly版本的
`pip install --pre torch torchvision torchaudio --index-url https://download.pytorch.org/whl/nightly/cpu
`
使用命令
`pip list | grep torch
`
看到类似这样的

torch                     2.3.0.dev20231224
torchaudio                2.2.0.dev20231224
torchvision               0.18.0.dev20231224

看到后面带dev的就可以下一步了

3、将/ChatGLM3/composite_demo/client.py
中的第140行左右
中的
device_map = “auto”
改为
device_map = “mps”


保存重新运行即可畅爽

**Reply by MoYJiang** [reply]
确实好了。质的飞跃。感谢！

**Reply by MoYJiang** [reply]
请问试过微调怎么样，目前微调怎么都不通，不论是否添加 .to('mps')或者 device_map=mps. 都会报错：
RuntimeError: Placeholder storage has not been allocated on MPS device!
另外还会有告警：[WARNING] NOTE: Redirects are currently not supported in Windows or MacOs.

**Reply by Ventery** [reply]
我也是刚刚学习这个，最近有点忙有空再研究研究。早知道这样就直接买主流硬件了

**Reply by MoYJiang** [reply]
调整了几个参数确实可调了（可以去看#253 torchrun命令，加--use_mps_device 1 参数），但是目前24g的m3够呛了，卡在这一步  “0%|          | 0/1000 [00:00<?, ?it/s] ”快半小时了没动，一开始报内存不够，后来环境变量PYTORCH_MPS_HIGH_WATERMARK_RATIO=0.0，倒是不报错了，gpu和mem都满了。。。MAC再等等吧，推理还是可以的，微调感觉要堆内存了。

**Reply by Ventery** [reply]
好

**Reply by jeceldx** [reply]
你好，请问 32g m1 pro内存只增不减，每次回答都会增加，这个问题您有遇到过吗？是否有办法在mps下量化的方法呢？

**Reply by Ventery** [reply]
遇到了，我还在看

**Reply by CJSen** [reply]
m2同样遇到内存问题了。正在找解决方案。

**Reply by Ventery** [reply]
这个只能减少每次history输入的条数了，限制为3条试试

**Reply by Ventery** [reply]
这个只能减少每次history输入的条数了，限制为3条试试

**Reply by southkorea2013** [reply]
请问这个内存占用不停增加的问题有解决方法吗？

**Reply by southkorea2013** [reply]
是PyTorch的问题，还是fschat调用的问题？

**Reply by Ventery** [reply]
内存本来就是随着对话次数在不断增加的，你可以限制一下每次输入历史的条数，在我的m2的机器上能把内存控制到一定程度不再增加的

**Reply by southkorea2013** [reply]
谢谢回复。我使用的是LangChain + Chatchat的项目，使用了chatglm3-6b+KB的方案。每次输入的历史条数为2，我发现几轮下来对应的运行chatglm3-6b的进程内存占用会飙升到70G左右，我用的是128G的M3。

**Comment by zRzRzRzRzRzRzR** [comment]
后续可能会有支持MLX的计划

**Comment by mobguang** [comment]
请问在Mac Studio M2 ultra上，可以用自己的dataset微调ChatGLM3吗，谢谢。

**Reply by zRzRzRzRzRzRzR** [reply]
可以

**Reply by mobguang** [reply]
谢谢回复。
请问有相关的指导文档吗，感谢。

**Reply by zRzRzRzRzRzRzR** [reply]
直接按照我们默认的文档就行了，但是估计你跳起来相当费劲，不知道你内存多少，mac至少要32G的内存才能调

没有专门为mac写的文档

**Reply by mobguang** [reply]
你提到“相当费劲”，请问指的是代码复杂度，还是说机器性能的限制，谢谢。

**Reply by zRzRzRzRzRzRzR** [reply]
机器性能，因为为微调占用显存，而你是mac，意味着是使用内存来完成这个任务，而且mac效率比英伟达显低太多了

**Reply by mobguang** [reply]
收到了，感谢！

**Comment by myEchos** [comment]
【咨询】Mac 上通过运行python3 web_demo_streamlit.py 会把模型下载到哪里？

我按照README.MD的步骤顺序往下执行，但是没有注意到我的Mac RAM 是不足以运行 chatglm 的。
但是此时已经完成了模型下载（7个 model-00001-of-00007.safetensors 大文件）。
运行过程中发现推理太慢（设备不符合条件），于是想要卸载，但是未能找到删除模型的方法，也未能找到模型所在的位置。

我做了一些尝试，但失败了：

尝试修改 “cli_demo.py”，使之在命令行输出
print(f"MODEL_PATH: {MODEL_PATH}")
print(f"TOKENIZER_PATH: {TOKENIZER_PATH}, os.path.abspath(TOKENIZER_PATH): {os.path.abspath(TOKENIZER_PATH)}")
但是得到的路径，里面并没有大文件。

通过 Mac 自带的存储文件清理，似乎我可控的文件夹内并没有大文件（难道是被统计到了“系统数据”中？）

尝试搜索，包括“THUDM/chatglm3-6b”、‘safetensors’，同样未能找到大文件。

希望能够得到各位的指导和帮助，先谢谢啦。

**Reply by southkorea2013** [reply]
如果是HuggingFace上的模型，一般在$HOME/.cache/huggingface/hub下面。

**Comment by ccall19** [comment]
mac studio 192G mps后端上进行Lora 微调，内存占用持续上涨直到突破物理内存，之后速度变得很慢，对比换到mac cpu 和 N卡上显存占用和速度都是稳定的，请问是否有解决方法？

**Reply by southkorea2013** [reply]
尝试用Apple自己的Framework MLX去跑LORA微调：
https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/LORA.md
不过现在MLX还不支持ChatGLM， Llama family 都支持。

**Comment by Zi-Gao** [comment]
实测在 macbook m3pro 18g 上，关闭出了终端的其他软件后，直接运行可以达到和 chatgpt3.5 甚至快一点的速度。

使用 chatglm-cpp 量化 fp16 模型，会显著降低回复延迟。若使用 int4 量化模型，可以达到10token/s

**Comment by WuXiaochengGG** [comment]
【咨询】
目前使用的是mac m1芯片。
试图本地部署ChatGLM3，按照教程走到pip install -r requirements.txt时出现问题。
依赖库vllm>=0.4.2无法安装。
找到vllm官方文档，其只支持linux系统。
请问该如何处理？

**Reply by HappyJimmyBoy** [reply]
我M3 Pro 也遇到这个问题了:AssertionError: vLLM only supports Linux platform (including WSL).

**Reply by jessienxx** [reply]
我用M2也遇到这个问题了，有人可以告知一下怎么解决嘛

**Reply by yuyunquan1** [reply]
我直接把那个依赖给注释掉了，其他的就正常安装，还真能用，cli和streamlit都可以，web_demo_gradio会报safetensors_rust.SafetensorError: Error while deserializing header: HeaderTooLarge的错误。

**Reply by monkeyin92** [reply]
me too! how to resolve

**Reply by SunshineRoc** [reply]
> 我直接把那个依赖给注释掉了，其他的就正常安装，还真能用，cli和streamlit都可以，web_demo_gradio会报safetensors_rust.SafetensorError: Error while deserializing header: HeaderTooLarge的错误。

其它能正常运行吗？

**Comment by iccolo** [comment]
【求助】使用mac m1 pro，按官方说明安装 PyTorch-Nightly成功，版本是 2.8.0.dev20250404，
pip3 install -r requirements.txt 安装依赖时报错：
  ERROR: Could not find a version that satisfies the requirement torch==2.5.1 (from versions: 2.6.0)
  ERROR: No matching distribution found for torch==2.5.1
